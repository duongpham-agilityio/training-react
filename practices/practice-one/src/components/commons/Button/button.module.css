.btn {
  display: flex;
  align-items: center;
  border: none;
  border-radius: 5px;
  list-style: none;
  cursor: pointer;
  font-family: var(--primary-font);
}

.btn:hover {
  opacity: 0.85;
}

.btn.btnDefault {
  background-color: silver;
}

.btn.btnPrimary {
  color: white;
  background-color: black;
}

.btn.btnSecondary {
  color: var(--primary-color);
  background-color: white;
}

.btn.btnSm {
  font-size: 11px;
  padding: 8px;
}

.btn.btnMd {
  font-size: 13px;
  padding: 10px;
}

.btn.btnLg {
  font-size: 15px;
  padding: 13px;
}

.btn.btnXl {
  font-size: 18px;
  padding: 20px;
}

.btn .btnIcon {
  font-size: inherit;
  padding: 0 5px;
}
