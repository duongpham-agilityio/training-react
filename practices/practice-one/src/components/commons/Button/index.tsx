import { ButtonHTMLAttributes, memo, ReactNode } from 'react';
import { Link, LinkProps } from 'react-router-dom';

// Styles
import styles from './button.module.css';

const selectTagName = {
  button: 'button',
  a: Link,
};

const selectSize = {
  sm: 'btnSm',
  md: 'btnMd',
  lg: 'btnLg',
  xl: 'btnXl',
};

const selectVariant = {
  default: 'btnDefault',
  primary: 'btnPrimary',
  secondary: 'btnSecondary',
  success: 'btnSuccess',
  danger: 'btnDanger',
  warning: 'btnWarning',
  info: 'btnInfo',
  light: 'btnLight',
  dark: 'btnDark',
};

type ButtonProps = ButtonHTMLAttributes<HTMLButtonElement> &
  Omit<LinkProps, 'to'> & {
    tag?: keyof typeof selectTagName;
    size?: keyof typeof selectSize;
    variant?: keyof typeof selectVariant;
    leftIcon?: ReactNode;
    rightIcon?: ReactNode;
    href?: string;
    description?: string;
  };

const Button = (props: ButtonProps) => {
  const {
    tag = 'button',
    size = 'md',
    variant = 'default',
    className = '',
    href = '',
    description = '',
    leftIcon,
    rightIcon,
    children,
    ...rest
  } = props;
  const TagName = selectTagName[tag];
  const classes = `${styles.btn} ${styles[selectVariant[variant]]} ${
    styles[selectSize[size]]
  } ${className}`;
  const attributes = {
    ...rest,
    to: href,
    className: classes,
  };

  return (
    <TagName {...attributes}>
      {leftIcon && <span className={styles.btnIcon}>{leftIcon}</span>}
      {children || description}
      {rightIcon && <span className={styles.btnIcon}>{rightIcon}</span>}
    </TagName>
  );
};

export default memo(Button);
