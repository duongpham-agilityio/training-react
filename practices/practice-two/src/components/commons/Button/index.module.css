.btn {
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 10px;
  font-weight: 500;
  line-height: 26.4px;
  text-transform: capitalize;
  border: none;
  padding: 10px;
}

.btn.w-sm {
  height: 48px;
  width: 48px;
}

.btn.w-lg {
  width: 100px;
  height: 46px;
}

.btn .icon {
  width: 14px;
}

.btn.default {
  background-color: var(--light-gray);
  color: white;
}

.btn.primary {
  background-color: var(--primary-color);
  color: white;
}

.btn.secondary {
  background-color: var(--secondary-color);
  color: black;
}

.btn.danger {
  background-color: var(--danger-color);
  color: white;
}

.btn.small {
  font-size: 14px;
}

.btn.medium {
  font-size: 16px;
}

.btn.large {
  font-size: large;
}

.btn.large {
  font-size: large;
}

.btn.b-md {
  border-radius: 13px;
}

.btn.b-lg {
  border-radius: 20px;
}

.btn:hover {
  opacity: 0.8;
  cursor: pointer;
}

/* Medium */
@media (min-width: 768px) {
  .btn.w-lg {
    min-width: 197px;
    height: 56px;
  }

  .btn.small {
    font-size: 20px;
  }

  .btn.medium {
    font-size: 22px;
  }

  .btn.large {
    font-size: large;
  }

  .btn.large {
    font-size: large;
  }

  .btn .icon {
    width: 22px;
  }
}
