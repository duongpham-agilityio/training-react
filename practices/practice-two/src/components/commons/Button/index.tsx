import { ButtonHTMLAttributes, memo } from 'react';

// Styles
import buttonStyles from 'components/commons/Button/index.module.css';

type Variant = 'default' | 'primary' | 'secondary' | 'danger';
type Size = 'small' | 'medium' | 'large';
type Border = 'b-md' | 'b-lg';
type Width = 'w-sm' | 'w-lg';
export type ButtonProps = ButtonHTMLAttributes<HTMLButtonElement> & {
  label?: string;
  leftIcon?: string;
  rightIcon?: string;
  variant?: Variant;
  border?: Border;
  size?: Size;
  width?: Width;
};

const Icon = ({
  icon,
  subTitle = 'icon action',
  ...rest
}: {
  icon?: string;
  subTitle?: string;
}) =>
  icon ? (
    <img
      src={icon}
      alt={subTitle}
      className={buttonStyles.icon}
      {...rest}
      width="22px"
      height="25px"
    />
  ) : null;

const ButtonIcon = memo(Icon);

const Button = (props: ButtonProps) => {
  const {
    label = '',
    leftIcon = '',
    rightIcon = '',
    className = '',
    variant = 'default',
    size = 'medium',
    border = 'b-md',
    width = 'w-sm',
    type = 'button',
    ...rest
  } = props;

  const classes = `${buttonStyles.btn} ${buttonStyles[variant]}
  ${buttonStyles[size]} ${buttonStyles[border]}
  ${buttonStyles[width]} ${className}`;

  return (
    <button className={classes} type={type} {...rest}>
      <ButtonIcon
        icon={leftIcon}
        subTitle="This is the left icon for the button"
      />

      {label}

      <ButtonIcon
        icon={rightIcon}
        subTitle="This is the right icon for the button"
      />
    </button>
  );
};

export default memo(Button);
