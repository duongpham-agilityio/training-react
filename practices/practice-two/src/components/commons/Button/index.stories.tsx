import { Meta, Story } from '@storybook/react';

// Component
import { Button, ButtonProps } from 'components/commons';

// Assets
import TrashIcon from 'assets/icons/trash.svg';

export default {
  title: 'Practice/components/commons/Button',
  component: Button,
} as Meta;

const Template: Story<ButtonProps> = (args) => <Button {...args} />;

export const Default = Template.bind({});
Default.args = {
  label: 'default',
  border: 'b-lg',
  size: 'small',
  width: 'w-lg',
};

export const Primary = Template.bind({});
Primary.args = {
  label: '1',
  size: 'small',
  variant: 'primary',
};

export const Secondary = Template.bind({});
Secondary.args = {
  label: 'secondary',
  size: 'medium',
  variant: 'secondary',
  border: 'b-lg',
  width: 'w-lg',
};

export const Danger = Template.bind({});
Danger.args = {
  label: '',
  leftIcon: TrashIcon,
  size: 'medium',
  variant: 'danger',
};
