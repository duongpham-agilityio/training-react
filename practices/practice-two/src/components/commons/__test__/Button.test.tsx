import { fireEvent, render } from '@testing-library/react';

import Button from 'components/commons/Button';

describe('Button', () => {
  it('match snapshot', () => {
    const { asFragment } = render(<Button label="click me" />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('render', () => {
    const { getByText } = render(<Button label="click me" />);

    expect(getByText(/click me/i)).toBeInTheDocument();
  });

  it('action', () => {
    const mockOnClick = jest.fn();
    const { getByText } = render(
      <Button label="click me" onClick={mockOnClick} />
    );
    const btn = getByText(/click me/i);

    fireEvent.click(btn);

    expect(mockOnClick).toHaveBeenCalled();
  });
});
