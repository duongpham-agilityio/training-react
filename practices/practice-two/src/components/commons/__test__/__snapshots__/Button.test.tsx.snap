// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button match snapshot 1`] = `
<DocumentFragment>
  <button
    class="undefined undefined
  undefined undefined
  undefined "
    type="button"
  >
    click me
  </button>
</DocumentFragment>
`;
